cng/des.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build windows
+// +build windows
+
+package cng
+
+import (
+	"crypto/cipher"
+	"runtime"
+
+	"github.com/microsoft/go-crypto-winnative/internal/bcrypt"
+	"github.com/microsoft/go-crypto-winnative/internal/subtle"
+)
+
+const desBlockSize = 8
+
+type desCipher struct {
+	kh bcrypt.KEY_HANDLE
+}
+
+func NewDESCipher(key []byte) (cipher.Block, error) {
+	kh, err := newCipherHandle(bcrypt.DES_ALGORITHM, "", key)
+	if err != nil {
+		return nil, err
+	}
+	c := &desCipher{kh: kh}
+	runtime.SetFinalizer(c, (*desCipher).finalize)
+	return c, nil
+}
+
+func NewTripleDESCipher(key []byte) (cipher.Block, error) {
+	kh, err := newCipherHandle(bcrypt.DES3_ALGORITHM, "", key)
+	if err != nil {
+		return nil, err
+	}
+	c := &desCipher{kh: kh}
+	runtime.SetFinalizer(c, (*desCipher).finalize)
+	return c, nil
+}
+
+func (c *desCipher) finalize() {
+	bcrypt.DestroyKey(c.kh)
+}
+
+func (c *desCipher) BlockSize() int { return desBlockSize }
+
+func (c *desCipher) Encrypt(dst, src []byte) {
+	if len(src) < desBlockSize {
+		panic("crypto/des: input not full block")
+	}
+	if len(dst) < desBlockSize {
+		panic("crypto/des: output not full block")
+	}
+	if subtle.InexactOverlap(dst, src) {
+		panic("crypto/des: invalid buffer overlap")
+	}
```

Comment:
I noticed the order of the checks is different than it is for `aesCipher`, so I looked at the standard library to see what it does--it's in this order for both aes and des.

But I also noticed that it only checks for overlap in the first block-length of each buffer:
https://github.com/golang/go/blob/795e779610cd1ef8ad4c137c067a23ace1190115/src/crypto/des/cipher.go#L42-L50
`alias.InexactOverlap(dst[:BlockSize], src[:BlockSize])`

I wrote a test to make sure I understood it right, and this fails on the branch right now (and doesn't if I swap in `des.NewCipher`):

```go
func TestSharedBufferWithoutLengthAdjustment(t *testing.T) {
	c := newCipher(encryptDESTests[0].key)
	srcPlusDst := make([]byte, c.BlockSize()*2)
	src := srcPlusDst
	dst := srcPlusDst[c.BlockSize():]
	c.Encrypt(dst, src)
}
```

I feel like the API would be better off (clearer) if it were stricter about this... but alas.

Reply:
Ah... but the CNG API fails if I just change the check to `subtle.InexactOverlap(dst[:desBlockSize], src[:desBlockSize])`. Adding `dst, src = dst[:desBlockSize], src[:desBlockSize]` before the check clears it up without changing the rest of the method, but I'm not sure if there's something deeper going on here.

Reply:
Filed https://github.com/golang-fips/openssl/issues/106 -- the logic is basically shared.

Reply:
Good catch. I've done some investigation and it is safe to trim `dst` and `src` to the block size, as DES encrypts/decrypts one full block at a time. This makes the behavior more similar to the standard library.